### PR TITLE
[ParseableInterface] Expose non-public types required for layout

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch does a pass over the module looking for stored properties
that contribute to the parent's storage but that use
non-public-or-usableFromInline types. It marks these types for printing,
while also hiding their non-required members.

For example, in a non-resilient module, a user may write this type:

```swift
internal class FileHandle {
  internal var fd: Int32
  init(fileDescriptor: Int32) { self.fd = fileDescriptor }
  deinit { close(fd) }
}

public struct File {
  internal var handle: FileHandle
}
```

Since `handle` contributes to the storage of the `File` type, we need to print it inside the declaration of `File`. Right now, we print the property as:

```swift
  internal var _: FileHandle
```
which signifies that the value was meant to be internal and not accessible from client code. But this caused an issue, because `FileHandle` is itself `internal`, and thus wouldn't be printed. Now, we find decls like FileHandle and elect to print their layout-required members as well. This example would be printed as:

```swift
internal class FileHandle {
  internal var _: Int32
}
public struct File {
  internal var _: FileHandle
}
```

Now, we could take this a step further and hide all the members of `FileHandle`, as we know it'll always be heap-allocated and pointer-sized, but that's not part of this patch.

It also splits off `ShouldPrintForParseableInterface` into its own file,
because it's grown significantly since it was first introduced in
`ASTPrinter`.

### CODE REVIEW

**Comment by brentdax** [comment]
Now that I’m seeing this written down, I wonder: Could you replace FileHandle with AnyObject? I’m not sure how that would work with non-class types—maybe you could declare it as a tuple of its member types, recursively until you hit public or builtin types? And I’m *definitely* not sure how it would work with existentials...

**Comment by harlanhaskins** [comment]
I wanted to avoid going down the rabbit hole of trying to synthesize layout-compatible types, electing for the simple thing of these partially-hidden stubs.

**Comment by jrose-apple** [comment]
I actually think we will be able to replace classes that way (though the extra inhabitants are different because AnyObject might be a tagged pointer and a pure Swift reference never will be), but we need basically this for structs and enums, so I think it's okay to move forward this way for now.

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -64,22 +65,7 @@ void PrintOptions::clearSynthesizedExtension() {
...
-static bool contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
-  auto *DC = ASD->getDeclContext()->getAsDecl();
-  if (!DC) return false;
-  auto *ND = dyn_cast<NominalTypeDecl>(DC);
-  if (!ND) return false;
-  return !ND->isResilient() && ASD->hasStorage() && !ASD->isStatic();
-}
-
-PrintOptions PrintOptions::printParseableInterfaceFile() {
+PrintOptions PrintOptions::printParseableInterfaceFile(ModuleDecl *module) {
```
Comment (jrose-apple) [inline lib/AST/ASTPrinter.cpp]: I'm honestly starting to think this doesn't belong on PrintOptions at all. Only the reusable or generic ones really deserve factory constructors; the rest should live at their use sites. That doesn't have to change in this PR, but it kind of lines up with the spiltting out to a separate file.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
+//=== ShouldPrintForParseableInterface.cpp - Parseable Interface filtering ===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: 2018

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+#include "ShouldPrintForParseableInterface.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/TypeWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/Type.h"
+
+using namespace swift;
+
+bool swift::isPublicOrUsableFromInline(const ValueDecl *VD) {
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: Why isn't this still `static`?

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  AccessScope scope =
+  VD->getFormalAccessScope(/*useDC*/nullptr,
+                           /*treatUsableFromInlineAsPublic*/true);
+  return scope.isPublic();
+}
+
+bool swift::contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: This one I think should probably just move onto AbstractStorageDecl, though maybe without the "isResilient" part. But that doesn't have to change now.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  public:
+    FindTypeDecls(TypeDecls &foundDecls): foundDecls(foundDecls) {}
+
+    void addIfNonPublic(TypeDecl *typeDecl) {
+      if (isPublicOrUsableFromInline(typeDecl))
+        return;
+      foundDecls.insert(typeDecl);
+    }
+
+    /// Finds all nominal types or typealiases referenced by the
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: Truncated comment.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        decl->getUnderlyingTypeLoc().getType().walk(*this);
+      }
+      return Action::Continue;
+    }
+  };
+  FindTypeDecls findTypeDecls;
+public:
+  FindReferencedNonPublicTypeDecls(TypeDecls &foundDecls):
+    findTypeDecls(foundDecls) {}
+  bool walkToDeclPre(Decl *decl) {
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: This should

- not walk into decls that aren't going to be printed
- not walk into function bodies

which makes me think you really want a worklist, not a single-pass walk. PrintAsObjC might be a good reference, since it has to do something similar for forward declarations.

File: test/ParseableInterface/non-public-required-types.swift [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+// RUN: %target-swift-frontend -emit-module -o %t/Test.swiftmodule %t.swiftinterface -disable-objc-attr-requires-foundation-module
+// RUN: %target-swift-frontend -emit-module -o /dev/null -merge-modules %t/Test.swiftmodule -module-name Test -emit-parseable-module-interface-path - | %FileCheck %s
+
+// RUN: %target-swift-frontend -emit-module -o %t/TestResilient.swiftmodule -enable-resilience %t.swiftinterface -disable-objc-attr-requires-foundation-module
+// RUN: %target-swift-frontend -emit-module -o /dev/null -merge-modules %t/TestResilient.swiftmodule -module-name TestResilient -enable-resilience -emit-parseable-module-interface-path - | %FileCheck %s
+
+
+// CHECK: internal struct GenericStruct<T> {
+internal struct GenericStruct<T> {
```
Comment (jrose-apple) [inline test/ParseableInterface/non-public-required-types.swift]: This logic'll need to handle enums too, but it might make sense to save that for another PR if it's not working yet.

**Review by jrose-apple — COMMENTED** [review]
I have some concerns, but this is a good start!

File: lib/AST/ShouldPrintForParseableInterface.cpp (hunk 3 above)
Comment (harlanhaskins) [reply lib/AST/ShouldPrintForParseableInterface.cpp]: It's used in a few places in `ASTPrinter` as well as this file.

**Comment by harlanhaskins** [comment]
@jrose-apple I've updated this PR to also support enums and protocols, and added them to the test.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 8]
```diff
@@ -50,25 +51,33 @@ bool swift::contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
...
 
     /// Finds all nominal types or typealiases referenced by the
     Action walkToTypePre(Type type) override {
       if (auto nominal =
           dyn_cast<NominalOrBoundGenericNominalType>(type.getPointer()))
         addIfNonPublic(nominal->getDecl());
+      if (auto enumType = dyn_cast<EnumType>(type.getPointer()))
+        addIfNonPublic(enumType->getDecl());
+      if (auto protocol = dyn_cast<ProtocolType>(type.getPointer()))
+        addIfNonPublic(protocol->getDecl());
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: These two should be redundant because they're both NominalOrBoundGenericNominalTypes. What's missing is protocol composition types, unless we treat that as having ProtocolTypes inside it?

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 9]
```diff
@@ -83,12 +92,63 @@ class FindReferencedNonPublicTypeDecls: public ASTWalker {
...
+      return false;
+    }
+
+    if (auto enumElement = dyn_cast<EnumElementDecl>(decl)) {
+      visitEnumElementDecl(enumElement);
+      return false;
+    }
+
+    if (auto protocol = dyn_cast<ProtocolDecl>(decl)) {
+      visitProtocolDecl(protocol);
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: I don't think the inheritance thing just applies to protocols. We also care about the layout of `@objc` enums, and if we're going to print classes we should probably print their superclass.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 10]
```diff
@@ -83,12 +92,63 @@ class FindReferencedNonPublicTypeDecls: public ASTWalker {
...
+  void visitAbstractStorageDecl(AbstractStorageDecl *decl) {
+    if (!contributesToParentTypeStorage(decl))
+      return;
+    decl->getInterfaceType().walk(findTypeDecls);
+  }
+
+  void visitEnumElementDecl(EnumElementDecl *decl) {
+    auto params = decl->getParameterList();
+    if (!params) return;
+    for (auto param : params->getArray()) {
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: ParameterList already has `begin`/`end`, so you should be able to drop the `getArray`.

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        addIfNonPublic(nominal->getDecl());
+      if (auto enumType = dyn_cast<EnumType>(type.getPointer()))
+        addIfNonPublic(enumType->getDecl());
+      if (auto alias = dyn_cast<NameAliasType>(type.getPointer())) {
+        // Add the typealias to the found decls
+        auto decl = alias->getDecl();
+        addIfNonPublic(decl);
+
+        // Also walk into the RHS of the typealias to find extra types we need.
+        decl->getUnderlyingTypeLoc().getType().walk(*this);
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: Shouldn't the visiting of the decl be responsible for that?

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    // If we see an enum element, walk into its associated values.
+    if (auto enumElement = dyn_cast<EnumElementDecl>(decl)) {
+      visitEnumElementDecl(enumElement);
+      return false;
+    }
+
+    // If we see a nominal type decl, walk its inherited types and generic
+    // requirements, and continue walking into it children.
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: Typo: "it children".

File: lib/AST/ShouldPrintForParseableInterface.cpp [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    auto params = decl->getParameterList();
+    if (!params) return;
+    for (auto param : *params) {
+      param->getInterfaceType().walk(findTypeDecls);
+    }
+  }
+
+  void visitNominalTypeDecl(NominalTypeDecl *decl) {
+    for (auto inherited : decl->getInherited())
+      inherited.getType().walk(findTypeDecls);
```
Comment (jrose-apple) [inline lib/AST/ShouldPrintForParseableInterface.cpp]: Uh, hm. If we don't print extensions, we might not get all the protocol conformances, which means we might not satisfy a generic constraint where this type is used. Is that a problem with our plan?

**Review by jrose-apple — COMMENTED** [review]
This seems nicely along the path we were going down but I have a new concern to work out (see below).

File: lib/AST/ShouldPrintForParseableInterface.cpp (hunk 13 above)
Comment (harlanhaskins) [reply lib/AST/ShouldPrintForParseableInterface.cpp]: Oh, yes that is a problem. Hmm...should I add extensions of this type to the printed set as well?

Comment (jrose-apple) [reply lib/AST/ShouldPrintForParseableInterface.cpp]: That might work for now. We probably do need to sit down and think about whether printing some kind of "layout token" would be easier all around, though.

Comment (harlanhaskins) [reply lib/AST/ShouldPrintForParseableInterface.cpp]: Or rather, walk declared conformances in extensions of this type, marking those for inclusion?

**Review by graydon — COMMENTED** [review]
Crude question: have you considered writing this as a request (in the request-evaluator sense)?

**Comment by harlanhaskins** [comment]
@graydon I think the request evaluator would be the wrong model here. Given that this essentially marks a type as required iff it's referenced in a layout-required context, you wouldn't be able to quickly ask the question of a given type. Structuring this as `(Type) -> Bool` would require walking all var decls every time you make the request, and caching those results. We could structure it as `(VarDecl) -> [Type]`, but then we get to a chicken-and-egg problem where we're only printing this VarDecl _after_ we have already skipped printing the required type.

Structuring this as a single walk where we start with all reachable top-level types and push discovered types onto a worklist seemed to be the right answer here.

**Comment by harlanhaskins** [comment]
Now, @jrose-apple and I discussed offline about the tradeoffs of this method vs. creating layout-compatible types.

We came up with a set of rules for non-public-or-`@usableFromInline` types (and please correct me if I'm wrong!):

| Type Kind | Strategy |
|-----------|----------|
| Typealiases | Recursively expand them in-place |
| Function types | Print their convention, and recursively expand parameters/return value |
| `@objc` classes or NSObject-derived classes | `AnyObject` |
| Pure Swift classes | `NativeObject` |
| Existential types | A tuple containing: `1 x Word` for value witness table, `N x Word` for protocol conformance witness tables, `1 x Any`/`1 x AnyObject` depending if the existential is class-bound |
| Enums | We need the layout, so print the original enum type |
| Structs | Recursively expand all fields into tuples |
| Metatypes | I think we can just use `Any.Type`? |

**Comment by jrose-apple** [comment]
> Enums: We need the layout, so print the original enum type

I think this is still going to have to be some kind of dummy type. "Print a fake enum with its payloads expanded."

**Comment by harlanhaskins** [comment]
Going to close this now, as we're going to take a different direction with this.
